### PR TITLE
Reducing State Accesses

### PR DESCRIPTION
**Pull request by sibanez12**
Initial attempt to reduce the memory bandwidth requirements for transport state variables. 

[Here is a google doc](https://docs.google.com/document/d/1ELDH_A03OfIeny9ZVj62oy5_gkIVF6VqwJKQj2QhmfQ/edit?usp=sharing) describing nanoPU state variable accesses.

    Attempted to consolidate a few state accesses across events.

    Refactored the creditEvent, deliveredEvent, and timeoutEvent.

    The deliveredEvent is now only invoked by ACKs.

    The creditToBtxEvent now performs RMW ops to update the credit
    state and the toBtx state. It also directly computes which
    pkts to transmit and inserts the tx_msg_id along with the tx_pkts
    bitmap into the scheduled_pkts_fifo. The scheduled_pkts_fifo
    has replaced the active_msgs_fifo.

    NOTE: if the scheduled_pkts_fifo is a PIFO this would enable some
    programability of scheduling policies when transmitting pkts.
    However, now that each entry contains multiple pkts, some flexibility
    is lost, but I think this should be sufficient - especially if the
    scheduling policy only depends on properties of the msg (as it does
    for Homa).

    The timeout event no longer needs to access the toBtx state, only
    the delivered state and the max_tx_pkt_offset state.

    The dequeue() operation in the Packetization module tracks the currently
    active tx_msg_id and tx_pkts and doesn't need to touch any other
    state.

### CODE REVIEW

File: nanoPU_sim.py [hunk 1]
```diff
@@ -217,6 +203,16 @@ def start(self):
                                            pkt[NDP].pkt_offset),
                             pkt[NDP].payload)
                     self.assemble_queue.put(data)
+                    pull_offset_diff = 1
```
Comment (sibanez12) [inline nanoPU_sim.py:206]: Just want to check: the credit is supposed to increase with _every_ DATA pkt and not every _new_ (i.e. unduplicated) DATA pkt?

Comment (serhatarslan-hub) [reply nanoPU_sim.py:206]: I am aware that it is a bit odd, but yes this was how the original NDP simulation was implemented. The idea is the following:
- If a data packet is received, it means the network can handle a new data packet now, so the receiver pulls a new data packet. 
- After all, if an old packet is being received for the second time, this means the sender timed out in the first time. This occurs when the ACK sent by the receiver regarding the first packet was lost which means the congestion is in the reverse path, so pulling a new packet is not harmful.

File: nanoPU_sim.py [hunk 2]
```diff
@@ -101,24 +103,14 @@ def compute_num_pkts(msg_len):
...
     # Prevent twos complement sign issue.
     return int(math.log(bitmap & -bitmap, 2))
 
-def priority_encoder(bitmap):
-    """Select an index based on the deployed priority selection policies
-    """
-    first_one = True
-    if first_one:
-        return find_first_one(bitmap)
-
```
Comment (serhatarslan-hub) [inline nanoPU_sim.py:121]: I had not deleted this function because different protocols (ie Homa) may need different priority encoding mechanism. Maybe I misunderstood the use of this function.

File: nanoPU_sim.py (hunk 1 above)
Comment (sibanez12) [reply nanoPU_sim.py:206]: Ah ok, thanks for that.

We don't need to add this to this PR, but I think it would be helpful to change the `getRxMsgID` extern call to `getRxMsgInfo` which returns: `rx_msg_id`, `is_new_msg`,  `is_new_pkt`, and `ack_no`. Where `is_new_pkt` indicates if this pkt has been received before. This would be useful for implementing Homa to allow the ingress pipeline to compute remaining msg size.

File: nanoPU_sim.py (hunk 2 above)
Comment (sibanez12) [reply nanoPU_sim.py:121]: We can add separate functions like `find_last_one` later if we need it.

File: nanoPU_sim.py (hunk 1 above)
Comment (serhatarslan-hub) [reply nanoPU_sim.py:206]: Agreed. Let's do that once we merge this PR. I can do that.

File: nanoPU_sim.py (hunk 2 above)
Comment (serhatarslan-hub) [reply nanoPU_sim.py:121]: Makes sense

File: nanoPU_sim.py [hunk 3]
```diff
@@ -101,24 +103,14 @@ def compute_num_pkts(msg_len):
     return msg_len/Simulator.max_pkt_len if (msg_len % Simulator.max_pkt_len == 0) else msg_len/Simulator.max_pkt_len + 1
 
 def find_first_one(bitmap):
-    """Find the first set bit in the provided bitmap
+    """Find the first set bit in the provided bitmap.
+       NOTE: this may not work as expected when doing: find_first_one(~value) if value is all 1's
```
Comment (sibanez12) [inline nanoPU_sim.py:107]: Might want to check this to make sure it is doing what you want.

Comment (serhatarslan-hub) [reply nanoPU_sim.py:107]: I have done some tests on edge cases I could think of (not the all ones case) and it worked.
The reason for me to change it into this was the fact that Python stores our bitmap variables as integers, so the most significant bit is used as the sign bit which is not displayed when we use bin() function to print or operate on the value. 
Please go ahead if you thinks something else works better here. I will also search for what we can do about it.

Comment (sibanez12) [reply nanoPU_sim.py:107]: Let's leave it as is for now.

File: nanoPU_sim.py [hunk 4]
```diff
@@ -435,33 +420,28 @@ def creditEvent(self, tx_msg_id, new_credit = None, dif_credit = None,
...
+                if relOp(compVal, cur_credit):
+                    self.credit[tx_msg_id] >>= new_credit
+                    self.log('Changing credit for msg {} from {} to {}'.format(tx_msg_id, cur_credit, self.credit[tx_msg_id]))
             else:
                 self.log('ERROR: creditEvent was triggered for unknown opCode: {}'.format(opCode))
 
-            if self.toBtx[tx_msg_id] & (1<<self.credit[tx_msg_id])-1 != 0:
-                self.enq_active_messages_fifo(tx_msg_id)
-        else:
-            self.log('ERROR: creditEvent was triggered for unknown tx_msg_id: {}'.format(tx_msg_id))
```
Comment (serhatarslan-hub) [inline nanoPU_sim.py:464]: Are we getting rid of the ERROR prompt here? (It is okay to delete it in terms of functionality but I think it would be nice to see detailed logs during experiments.)

Comment (sibanez12) [reply nanoPU_sim.py:464]: The `creditToBtxEvent` is now invoked by both NACKs and PULLs. When a NACK invokes this action it will set `new_credit = None`.

Comment (serhatarslan-hub) [reply nanoPU_sim.py:464]: Right. Thanks. This else clause was for unknown tx_msg_id values. Maybe we could ad it as a separate if statement to check that. Leaving this for now. :)

File: nanoPU_sim.py [hunk 5]
```diff
@@ -435,33 +420,28 @@ def creditEvent(self, tx_msg_id, new_credit = None, dif_credit = None,
...
+            # NOTE: here's another Read-Modify-Write to update toBtx state, which
+            #   should be combined with the above operation in the actual HW implementation
+            tx_pkts = self.toBtx[tx_msg_id] & (1<<self.credit[tx_msg_id])-1
+            if tx_pkts != 0:
+                # schedule the pkts for transmission
+                self.scheduled_pkts_fifo.put((tx_msg_id, tx_pkts))
```
Comment (serhatarslan-hub) [inline nanoPU_sim.py:442]: I should admit that this is a clever idea. For the paper we write in the future, we should add the following disclaimer into the paper though:
> When the msg is pushed into the scheduled_pkts_fifo, the toBtx packets are locked for this iteration of the message. By the time the message is dequeued, the message may have additional packets that also have have toBtx bits set. Although those additional packets might have lower sequence number (higher priority for transmission) they are not transmitted until the scheduled_pkts_fifo re-enqueues this message. This may slightly alter the arrival order of packets, but this alteration would be limited to a small number of packets. (highest alteration amount is RTTbytes -1 packets for NDP)

File: nanoPU_sim.py [hunk 6]
```diff
@@ -478,13 +458,11 @@ def timeoutEvent(self, tx_msg_id, rtx_offset):
...
             self.log('Pkts to retransmit: {:b}'.format(rtx_pkts))
-            toBtx_bitmap = self.toBtx[tx_msg_id]
-            self.toBtx[tx_msg_id] = toBtx_bitmap | rtx_pkts
             # reschedule timer for this msg
             self.rescheduleTimerEvent(tx_msg_id, self.max_tx_pkt_offset[tx_msg_id])
             if rtx_pkts != 0:
-                # make the message active
-                self.enq_active_messages_fifo(tx_msg_id)
+                # schedule the pkts for transmission
+                self.scheduled_pkts_fifo.put((tx_msg_id, rtx_pkts))
```
Comment (serhatarslan-hub) [inline nanoPU_sim.py:465]: Does this mean the scheduled_pkts_fifo can have two entries that belong to the same msg (one inserted by the creditToBTx event and one inserted by the Timeout event)

File: nanoPU_sim.py (hunk 5 above)
Comment (sibanez12) [reply nanoPU_sim.py:442]: Is the situation you are concerned about something like this: a PULL pkts arrives and a new pkt is scheduled for tx (e.g. pkt 6) then immediately after the PULL pkt, a NACK arrives indicating that pkt 4 should be retransmitted. In which case, pkt 6 is transmitted before pkt 4?

File: nanoPU_sim.py (hunk 6 above)
Comment (sibanez12) [reply nanoPU_sim.py:465]: Yes, that's right. Both the timeout event and the creditToBtx event can schedule pkts for transmission. They should be scheduling different pkts within the msg.

File: nanoPU_sim.py [hunk 7]
```diff
@@ -519,49 +497,37 @@ def start(self):
...
-        # mark msg as inactive (flip bit)
-        self.active_messages_bitmap = self.active_messages_bitmap ^ (1<<tx_msg_id)
-        # lookup which pkts of this msg need to be transmitted
-        toBtx_bitmap = self.toBtx[tx_msg_id]
-        credit = self.credit[tx_msg_id]
-        credit_bitmap = (1<<credit)-1
-        # the only pkts that are eligible are the ones at an index < credit
-        eligible_pkts_bitmap = toBtx_bitmap & credit_bitmap
+        # check if we've finished sending all the currently active pkts
+        if self.active_tx_pkts == 0:
```
Comment (serhatarslan-hub) [inline nanoPU_sim.py:517]: We are transferring all the permitted packets of a message before handling another message. This is definitely a legitimate option. However this means that two messages become ready to send 10 packets at the same time, one of them send its 10 packets in a single burst and the other waits for 10 packet serialization times before being able to send any packet. I think this increases burstiness and promotes more congestion in the network.
Especially for nanoPU world where most of the messages are short so that they would be sent as bursts (creating incast) the natural pacing (msg handling in round robin fashion) could be somewhat helpful. What do you think?

File: nanoPU_sim.py (hunk 5 above)
Comment (serhatarslan-hub) [reply nanoPU_sim.py:442]: Yep

File: nanoPU_sim.py (hunk 6 above)
Comment (serhatarslan-hub) [reply nanoPU_sim.py:465]: awesome! Thanks for clarification.

File: nanoPU_sim.py (hunk 7 above)
Comment (sibanez12) [reply nanoPU_sim.py:517]: In what situations would something like this happen? i.e. a PULL pkt arrives and makes a whole bunch of pkts available for tx? Maybe if the PULLs are significantly reordered or some are dropped in the network? In the usual case each PULL should only trigger generation of 1 pkt, right?

File: nanoPU_sim.py (hunk 5 above)
Comment (sibanez12) [reply nanoPU_sim.py:442]: Yeah this might cause some re-ordering of generated DATA pkts when we compare to the actual NDP implementation. Something to keep in mind.

File: nanoPU_sim.py (hunk 7 above)
Comment (serhatarslan-hub) [reply nanoPU_sim.py:517]: The examples you give are possible. 
Yes ideally a PULL should generate only a single packet but this is not guaranteed and not rare enough because of reordering.
In addition, imagine CPU writes 2 messages to the packetization buffer back to back. Let's say by the time the CPU completes writing the second message, the first messages has not been dequed to the egress pipeline yet because of already existing messages in the tx_fifo. Then the second message will not be able to send any packets at all until the first message sends out 10 packets into the network.

**Review by serhatarslan-hub — APPROVED** [review]
Thanks for the state updates. I think those changes would not hurt NDP behavior in general, except couple of small notes I commented during the review. 
We should continue thinking about the burstiness of messages. (All ready packets of the same message are transmitted at the same time before handling a different message rather than sending 1 packet of a message at a time)
However we should be aware of the fact that the order of transmitted packets and timeout mechanism is going to be different then the simulators we are going to compare ourselves to. This will make it difficult to verify our protocol implementations. I think it should be sufficient to verify reliable delivery + credit alteration mechanism with small tests.

File: nanoPU_sim.py (hunk 7 above)
Comment (sibanez12) [reply nanoPU_sim.py:517]: I agree PULL pkts may generate some bursts of multiple pkts from the same msg. However, the way that this behavioral model implements the scheduling of tx pkts for msg writes from the CPU does not accurately represent the HW implementation. After the CPU finishes writing each pkt, the msg processing thread will insert an entry into the scheduled_pkts_fifo. The creditToBtxEvent thread may insert entries between these writes, so there won't necessarily be a burst of pkts from the same msg.

I don't think it's worth updating this behavioral model to be exactly like the HW implementation, there are diminishing returns.

Another edge case that comes up in the HW implementation but not here in the behavioral model is: what happens when a PULL pkt tries to pull a pkt that has not been written by the CPU yet? I think we may need another state variable (e.g. `available`) so that the creditToBtxEvent doesn't try to send an unavailable pkt...

Comment (serhatarslan-hub) [reply nanoPU_sim.py:517]: Alright. Makes sense. 
Let's continue thinking about bursty scenarios.
